pingora/examples/client.rs

```diff
@@ -12,29 +12,58 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use anyhow::Error;
 use pingora::connectors::http::Connector;
 use pingora::upstreams::peer::HttpPeer;
 use pingora_http::RequestHeader;
+use regex::Regex;
 
 #[tokio::main]
-async fn main() {
+async fn main() -> Result<(), Error> {
     let connector = Connector::new(None);
 
-    let mut peer = HttpPeer::new("1.1.1.1:443", true, "one.one.one.one".into());
+    // create the HTTP session
+    let peer_addr = "1.1.1.1:443";
+    let mut peer = HttpPeer::new(peer_addr, true, "one.one.one.one".into());
     peer.options.set_http_version(2, 1);
-    let (mut http, _reused) = connector.get_http_session(&peer).await.unwrap();
-
-    let mut new_request = RequestHeader::build("GET", b"/", None).unwrap();
-    new_request
-        .insert_header("Host", "one.one.one.one")
-        .unwrap();
-    http.write_request_header(Box::new(new_request))
-        .await
-        .unwrap();
+    let (mut http, _reused) = connector.get_http_session(&peer).await?;
+
+    // perform a GET request
+    let mut new_request = RequestHeader::build("GET", b"/", None)?;
+    new_request.insert_header("Host", "one.one.one.one")?;
+    http.write_request_header(Box::new(new_request)).await?;
+
     // Servers usually don't respond until the full request body is read.
-    http.finish_request_body().await.unwrap();
-    http.read_response_header().await.unwrap();
-    println!("{:#?}", http.response_header().unwrap());
-    // TODO: continue reading the body
-    // TODO: return the connection back to the `connector` (or discard it)
+    http.finish_request_body().await?;
+    http.read_response_header().await?;
+
+    // display the headers from the response
+    if let Some(header) = http.response_header() {
+        println!("{:#?}", header);
+    } else {
+        return Err(anyhow::anyhow!("No response header"));
+    };
+
+    // collect the response body
+    let mut response_body = String::new();
+    while let Some(chunk) = http.read_response_body().await? {
+        response_body.push_str(&String::from_utf8_lossy(&chunk));
+    }
+
+    // verify that the response body is valid HTML by displaying the page <title>
+    if let Some(title) = Regex::new(r"<title>(.*?)</title>")?
```

Comment:
A convenient way to wrap a Pingora error around the causing `Regex` error:
```
if let Some(title) = Regex::new(r"<title>(.*?)</title>")
    .or_err(InternalError, "Failed to compile regex")?
    ...
```

Reply:
:+1: 5833556e5f8d9ccb346ef586cb1b5f51add64357